mpi4jax/_src/jax_compat.py

```diff
@@ -59,8 +76,34 @@ def check_jax_version():
     from jax.interpreters.mlir import token_type  # noqa: F401
 
 
+# TODO: remove this code once we only support jax > 0.4.14
+if versiontuple(jax.__version__) >= (0, 4, 14):
+    from jax.core import ShapedArray  # noqa: F401
+else:
+    from jax.abstract_arrays import ShapedArray  # noqa: F401
+
+
+# TODO: remove this code once we only support jax > 0.4.14
+if versiontuple(jax.__version__) >= (0, 4, 16):
+
+    def register_effect(EffectType):
+        from jax.interpreters import mlir
+        from jax._src.effects import (
+            control_flow_allowed_effects,
+            custom_derivatives_allowed_effects,
+        )
+
+        effect = EffectType()
+        mlir.lowerable_effects.add_type(EffectType)
+        control_flow_allowed_effects.add_type(EffectType)
+        # Effects must be added to the allow_effects list in order to work within
+        # custom_vjp. See google/jax#11916
+        custom_derivatives_allowed_effects.add_type(EffectType)
```

Comment:
I see some more effects here:

https://github.com/google/jax/blob/7bc01d9472fbe872e14db59eed81a26b0e8c078b/jax/_src/debugging.py#L68-L76

Namely, `effects.ordered_effects`, `effects.lowerable_effects` (is this an alias for `mlir.lowerable_effects`?), `effects.remat_allowed_effects`. Especially `ordered_effects` sounds relevant for us?!

Reply:
I simply adapted the code following what was moved around. I suspect we might tag this as being allowed under remat, but would rather add a test later on than quickly add it here and risk getting segfaults.
In any case, remat is not supported even now.

Reply:
The module docstring here seems to imply that we can eliminate tokens if we use `ordered_effects`: https://github.com/google/jax/blob/main/jax/_src/effects.py

Big if true.

Reply:
Also, [`mlir.lowerable_effects` *is* an alias for `effects.lowerable_effects`](https://github.com/google/jax/blob/main/jax/_src/interpreters/mlir.py#L64), so we should be using that instead.

Reply:
_![LiesCalcAndCompTold.p_t_df](https://githu_[]()_____b.com/mpi4jax/mpi4jax/files/12671790/LiesCalcAndCompTold.pdf)
e
[]()[]()_